rfcs/0010-multi-tenant-workload-identity/README.md

```diff
@@ -0,0 +1,1251 @@
+
+In sight of the technical background presented above, our proposal becomes simpler.
+The only solution to support multi-tenant workload identity at the object-level for
+the Flux APIs is to associate the Flux objects with Kubernetes `ServiceAccounts`.
+We propose to build the `ServiceAccount` token creation and exchange logic into
+the Flux controllers through a library in the `github.com/fluxcd/pkg` repository.
+
+### API Changes
+
+For all the Flux APIs interacting with cloud providers (except `Kustomization`,
+see the paragraph below), we propose introducing the field `spec.serviceAccountName`
+(if not already present) for specifying the Kubernetes `ServiceAccount` on the same
+namespace of the object that must be used for getting access to the respective cloud
+resources. This field would be optional, and when not present the original behavior
+would be observed, i.e. the feature only activates when the field is present and a
+cloud provider among `aws`, `azure` or `gcp` is specified in the `spec.provider`
+field. So if only the `spec.provider` field is present and set to a cloud provider,
+then the controller would use single-tenant workload identity as it would prior to
+the implementation of this RFC, i.e. it would use its own identity for the operation.
+
+Note that this RFC does not seek to change the behavior when `spec.provider` is set
+to `generic` (or left empty, when it defaults to `generic`), in which case the field
+`spec.secretRef` can be used for specifying the Kubernetes `Secret` containing the
+credentials (or `spec.serviceAccountName` in the case of the APIs dealing with
+container registries, through the `imagePullSecrets` field of the `ServiceAccount`).
+
+The `Kustomization` API uses Key Management Services (KMS) for decrypting
+SOPS-encrypted secrets. We propose adding the dedicated optional field
+`spec.decryption.serviceAccountName` for multi-tenant workload identity
+when intercting with the KMS service. We choose having a dedicated field
+for the `Kustomization` API because the field `spec.serviceAccountName`
+already exists and is used for a major part of the functionality which
+is authenticating with the Kubernetes API when applying resources. If
+we used the same field for both purposes users would be forced to use
+multi-tenancy for both cloud and Kubernetes API interactions.
+
+In the `Bucket` API we have previously introduced the optional field `spec.sts`.
+This field has configuration for interacting with the STS service of the cloud
+provider, most specifically the `spec.sts.endpoint` field. In the specific case
+of the `Bucket` API, there's also the field `spec.sts.provider`, which allows
+using other STS configurations relavant only to the `generic` bucket provider,
+whose underlying implementation uses the MinIO SDK (the AWS S3-compatible object
+storage server). The same does not apply to the other Flux APIs, so we propose
+introducing only a single optional field `spec.stsEndpoint` instead.
```

Comment:
@stefanprodan please review this part, I think `spec.stsEndpoint` is better for the other APIs, `spec.sts` makes sense only for Bucket

Reply:
(By the way, the only provider that supports this is AWS. Azure and GCP don't support custom STS endpoints.)

Reply:
Ok let's not add the sts setting in here. Let's see if anyone needs it after we implement the SA.

Reply:
We don't need to implement this until somebody asks for it, but since we already have `spec.sts.endpoint` for the Bucket API anyway maybe we should at least agree on a plan here for when somebody asks for it, so I updated the RFC with what we decided here. Please let me know if you prefer otherwise (removing this paragraph about STS endpoint altogether and the update I just added to the RFC)

Reply:
Paragraph removed as discussed via DM :ok_hand: